File: src/Julia/Duplicate-Encoder.jl

using DataStructures, Test


"""
出现多于1次的字符编码为右括号，只有1次的为左括号
"""
function duplicateencode(word::String)::String
    lower_word = lowercase(word)

    dict = counter(lower_word)
    encode(c::Char)::Char = dict[c] > 1 ? ')' : '('

    map(encode, lower_word)
end

@test duplicateencode("Success") == ")())())"

File: src/JavaScript/Duplicate-Encoder.js
const counter = require("./toolkit/Vector").counter;

/**
 * 重复字符编码器
 * @param {String} word 
 */
function duplicateEncode(word) {
  const lowerWord = word.toLowerCase();
  const dict = counter(lowerWord);
  return lowerWord.split("").map(c => dict[c] === 1 ? "(" : ")").join("");
}

console.log(duplicateEncode("Success"));